nb_conda_kernels/tests/test_notebook.py

```diff
@@ -0,0 +1,173 @@
+            'IPYTHONDIR': self.ipydir.name,
+        }
+        self.dirs.append(self.ipydir)
+        self.dirs.append(self.home)
+        self.dirs.append(self.config_dir)
+        self.dirs.append(self.nbdir)
+        os.makedirs(os.path.join(self.nbdir.name,
+                                 os.path.join(u'sub dir1', u'sub dir 1a')))
+        os.makedirs(os.path.join(self.nbdir.name,
+                                 os.path.join(u'sub dir2', u'sub dir 1b')))
+
+        if self.xunit:
+            self.add_xunit()
+
+        # If a url was specified, use that for the testing.
+        if self.url:
+            try:
+                alive = jstest.requests.get(self.url).status_code == 200
+            except:
+                alive = False
+
+            if alive:
+                self.cmd.append("--url=%s" % self.url)
+            else:
+                raise Exception('Could not reach "%s".' % self.url)
+        else:
+            # start the ipython notebook, so we get the port number
+            self.server_port = 0
+            self._init_server()
+            if self.server_port:
+                self.cmd.append('--url=http://localhost:%i%s' % (
+                    self.server_port, self.base_url))
+            else:
+                # don't launch tests if the server didn't start
+                self.cmd = [
+                    jstest.sys.executable, '-c', 'raise SystemExit(1)']
+
+    def add_xunit(self):
+        """ Hack the setup in the middle (after paths, before server)
+        """
+        super(NBCondaKernelsTestController, self).add_xunit()
+
+        # ensure the system-of-interest is installed and enabled!
+        with patch.dict(os.environ, self.env):
+            nb_conda_kernels.install.install(
+                prefix=os.environ.get("CONDA_ENV_PATH",
+                                      os.environ.get("CONDA_DEFAULT_ENV")),
+                enable=True,
+                verbose=True)
+
+    def cleanup(self):
+        if hasattr(self, "stream_capturer"):
+            captured = self.stream_capturer.get_buffer().decode(
+                'utf-8', 'replace')
+            with open(TEST_LOG, "a+") as fp:
+                fp.write("-----------------------\n{} results:\n{}\n".format(
+                    self.section,
+                    self.server_command))
+                fp.write(captured)
+
+        super(NBCondaKernelsTestController, self).cleanup()
+
+
+def prepare_controllers(options):
+    """Monkeypatched prepare_controllers for running widget js tests
+
+    instead of notebook js tests
+    """
+    if options.testgroups:
+        groups = options.testgroups
+    else:
+        groups = ['']
+    return [NBCondaKernelsTestController(g, extra_args=options.extra_args)
+            for g in groups], []
+
+
+def test_notebook():
+    with patch.object(jstest, 'prepare_controllers', prepare_controllers):
+        jstest.main()
+
+
+if __name__ == '__main__':
+    test_notebook()
```

Comment:
Yeah. I'm tired of building this same infrastructure over and over with slight differences, but it works (mostly) reliably all over (docker, cloud, windows) which makes it worthwhile.

Having all of this packed up into a single package, only required at build time, would be super helpful, but I'd have to look at all the crazy (like nb_anacondacloud) that is needed to do different kinds of mocks. Too much mocking/spying is bad in the notebook, as it gets much harder to tell how unrelated things might work together.